File: bzrlib/textfile.py
from itertools import chain 

from bzrlib.errors import BinaryFile
from bzrlib.iterablefile import IterableFile
from bzrlib.osutils import file_iterator
def text_file(input):
    """Produce a file iterator that is guaranteed to be text, without seeking.
    BinaryFile is raised if the file contains a NUL in the first 1024 bytes.
    """
    first_chunk = input.read(1024)
    if '\x00' in first_chunk:
        raise BinaryFile()
    return IterableFile(chain((first_chunk,), file_iterator(input)))


def check_text_lines(lines):
    """Raise BinaryFile if the supplied lines contain NULs.
    Only the first 1024 characters are checked.
    """
    f = IterableFile(lines)
    if '\x00' in f.read(1024):
        raise BinaryFile()

File: crates/osutils/src/textfile.rs
use std::fs::File;
use std::io::{Error, Read};
use std::path::Path;

/// Return false if the supplied lines contain NULs.
///
/// Only the first 1024 characters are checked.
pub fn check_text_lines<I>(lines: I) -> bool
where
    I: IntoIterator<Item = Vec<u8>>,
{
    let mut buffer = [0u8; 1024];
    let mut offset = 0;
    for line in lines.into_iter() {
        if line.iter().any(|&c| c == 0) {
            return false;
        }
        if offset + line.len() > 1024 {
            break;
        }
        buffer[offset..offset + line.len()].copy_from_slice(&line);
        offset += line.len();
    }
    if buffer[..offset].iter().any(|&c| c == 0) {
        return false;
    }
    true
}

/// Check whether the supplied path is a text, not binary file.
///
/// Raise BinaryFile if a NUL occurs in the first 1024 bytes.
pub fn check_text_path<P: AsRef<Path>>(path: P) -> Result<bool, Error> {
    let file = File::open(path)?;
    let mut buffer = Vec::new();
    let mut handle = file.take(1024);
    handle.read_to_end(&mut buffer)?;

    Ok(buffer.iter().all(|&byte| byte != 0))
}
